# Prediction of financial distress

## Description

This dataset (taken from [here](https://www.kaggle.com/shebrahimi/financial-distress)) contains data temporal data abount several hundreds of companies. There are ~80 features for each company times a selected window with one target variable: the level of "financial distress" the company currently faces.

## Task

Predict whether a company faces financial distress based on the current and past input features.

## Dataset columns description

```
Company            - Company id
Time               - Timestamp id for each company; varies between 1 to 14.
Financial Distress - The target variable
x1-x80             - Financial and non-financial characteristics for ech company.
```

Notes from authors:
- If the target "Financial Distress" variable if it is greater than -0.50 the company should be considered healthy. Otherwise, it would be regarded financially distressed
- Input features belong to the previous time period, which should be used to predict whether the company will be financially distressed or not
- Feature x80 is a categorical variable
- The data set is imbalanced. There are 136 financially distressed companies against 286 healthy ones i.e.136 firm-year observations are financially distressed while 3546 firm-year observations are healthy.


Other Notes:
- 1 timestep = 1 year (authors refer to firm-year observations)
- All "positive" time series (companies that are financially distressed) ends with the target variable below the threshold. There are no companies that becomes "healthy" again. 
- Also all positive samples ends after one year in financial distress. Once the target variable becommes less then 0.5, it ends.
- The above points leads to a hypothesis that the dataset is actually about companies going bankrupt or defaulted loans.

## Imports & Drive mount

In [ ]:
%tensorflow_version 2.x
%matplotlib inline

import random

import numpy as np
import pandas as pd

import matplotlib
import matplotlib.pyplot as plt

from tqdm.notebook import tqdm

In [ ]:
matplotlib.rcParams['figure.figsize'] = (15, 6)
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


## Data loading and basic analysis

In [ ]:
df = pd.read_csv('/content/drive/My Drive/ml-college/time-series-analysis/data/Financial Distress.csv')
df

Company  Time  Financial Distress      x1  ...  x80       x81  x82  x83
0           1     1            0.010636  1.2810  ...   22  0.060390   30   49
1           1     2           -0.455970  1.2700  ...   22  0.010636   31   50
2           1     3           -0.325390  1.0529  ...   22 -0.455970   32   51
3           1     4           -0.566570  1.1131  ...   22 -0.325390   33   52
4           2     1            1.357300  1.0623  ...   29  1.251000    7   27
...       ...   ...                 ...     ...  ...  ...       ...  ...  ...
3667      422    10            0.438020  2.2605  ...   37  0.436380    4   41
3668      422    11            0.482410  1.9615  ...   37  0.438020    5   42
3669      422    12            0.500770  1.7099  ...   37  0.482410    6   43
3670      422    13            0.611030  1.5590  ...   37  0.500770    7   44
3671      422    14            0.518650  1.6148  ...   37  0.611030    8   45

[3672 rows x 86 columns]

In [ ]:
df.describe()

Company         Time  ...          x82          x83
count  3672.000000  3672.000000  ...  3672.000000  3672.000000
mean    182.084423     7.528322  ...    13.122277    33.044935
std     117.024636     4.064016  ...     9.465907    13.714563
min       1.000000     1.000000  ...     1.000000     2.000000
25%      80.000000     4.000000  ...     6.000000    21.000000
50%     168.000000     7.000000  ...    11.000000    34.000000
75%     264.250000    11.000000  ...    17.000000    44.000000
max     422.000000    14.000000  ...    49.000000    74.000000

[8 rows x 86 columns]

^ From the data we can see that:
- There are 422 companies in the dataset
- There are max 14 time steps as advertised
- Some of the features contains huge outliers (x8, x12, ..)
- Besides x80, some other features seems to be rather categorical (x72, x79, ..)